### ISSUE TITLE
Please provide a way to disable the DeveloperExceptionPage during development.

### ISSUE THREAD

**Issue by aetos382** [dotnet/aspnetcore#56906]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

If I am using the `WebApplication.Create` method, the DeveloperExceptionPage is automatically enabled in the development environment.
Please provide a way to disable this.

Similar requests have been submitted and rejected in the past. (#43870, #47293)

I want to observe the behavior when an exception occurs in the production environment during development.
But when DeveloperExceptionPage is enabled, a user-friendly display is shown, and I don't get the actual error that would be returned in the production environment (e.g. `application/problem+json`).

Solutions such as setting the DOTNET_ENVIRONMENT environment variable to Production or using WebApplication.CreateSlimBuilder are difficult to use.

If the `DOTNET_ENVIRONMENT` environment variable is changed, the `appsettings.*.json` file that is loaded will change, which may result in settings that are unsuitable for the development environment, for example, preventing database connections.

And if I use the `WebApplication.CreateSlimBuilder` method instead of the `WebApplication.Create` method, there are some services that will be disabled, so I have to investigate and supplement them.
I only want to disable the DeveloperExceptionPage, and I don't want to disable any other services.

I don't want to complete and deploy this application with DeveloperExceptionPage disabled. I only want to disable it temporarily while developing in a local environment.
The method mentioned above is too costly for this purpose. I only want to temporarily disable DeveloperExceptionPage with minimal changes.

### Describe the solution you'd like

There are multiple ways to do this. Any method is fine.

For example:

```cs
services.DisableDeveloperExceptionPage();
```

Or

```cs
WebApplication.CreateBuilder(
  new WebApplicationOptions {
    Args = args,
    DisableDeveloperExceptionPage = true
  });
```

**Comment by aetos382** [comment]
I will now give you a more specific example of something I recently encountered.

I recently upgraded my application from .NET 7.0 to .NET 8.0.

When using .NET 7.0, I was setting the `JsonTypeInfoResolver` as follows to accelerate the (de)serialization of API request/response payloads.

```cs
services
    .AddControllers()
    .AddJsonOptions(static options =>
    {
        options.JsonSerializerOptions.TypeInfoResolver = JsonTypeInfoResolver.Combine(
            MyApiPayloadsSerializerContext.Default,
            new DefaultJsonTypeInfoResolver());
    });
```

It is important to combine `MyApiPayloadsSerializerContext.Default` with `new DefaultJsonTypeInfoResolver()`.
If I only set `MyApiPayloadsSerializerContext.Default` as follows, it will fail to serialize objects that it does not support (e.g. `ProblemDetails`) if they are returned.

```cs
services
    .AddControllers()
    .AddJsonOptions(static options =>
    {
        options.JsonSerializerOptions.TypeInfoResolver = MyApiPayloadsSerializerContext.Default;
    });
```

When writing as above, if the application receives an invalid request, the framework will try to return `ProblemDetails` to indicate that the request is invalid, but an error will occur that it cannot be serialized, so the original error will be masked.

In .NET 8.0, it would be better to rewrite it as follows.

```cs
services
    .AddControllers()
    .AddJsonOptions(static options =>
    {
        options.JsonSerializerOptions.TypeInfoResolverChain.Insert(0, MyApiPayloadsSerializerContext.Default);
    });
```

The `TypeInfoResolverChain` includes the reflection-based `DefaultJsonTypeInfoResolver` by default, so there is no need to explicitly add it as before.

I wanted to verify that my application was not broken by this change.
In other words, I wanted to make sure that if I sent an invalid request to the application, I would receive a response in the `application/problem+json` format that correctly indicated that the request was invalid, rather than an error that `ProblemDetails` could not be serialized.

However, if DeveloperExceptionPage is enabled, even if I send a request using `curl`, the console displays the error message generated by DeveloperExceptionPage, rather than the response in `application/problem+json` format.

I just want to confirm it, but if I change the `DOTNET_ENVIRONMENT` environment variable or use the `WebApplication.CreateSlimBuilder` method, a lot of other changes will be required.
So, without doing those things, I just want to disable the DeveloperExceptionPage.

**Comment by amcasey** [comment]
Thanks for the detailed explanation of your scenario and motivation!

**Comment by davidfowl** [comment]
c@captainsafia have we seen other requests for this?

**Comment by captainsafia** [comment]
> [@captainsafia](https://github.com/captainsafia) have we seen other requests for this?

Outside of the ones linked in the original issue, there are semi-related ones like https://github.com/dotnet/aspnetcore/issues/52622.

It seems like the overall goal people have here is to be able to model behavior that happens in production environments when exceptions occur locally.

I think the "right" thing to do here is to use a non-Development environment if you want to replicate Production-behavior locally but there definitely is a cliff you follow of because this will disable other features that are behind a Development check, like validating scopes on services in the DI container or injection of user secrets.

Adding `DisableDeveloperExceptionPage` to `WebApplicationOptions` could help here but given how targeted the API is I don't see overwhelming demand for it.
